feat(minhash): optional CuPy backend for MinHash.update_batch

dipeshbabu opened the pull request:
Add an opt-in GPU backend for MinHash.update_batch using CuPy.

- Hashing and permutation generation remain on CPU and unchanged.
- When enable_gpu() is called and CuPy is available, the internal
  (hv * a + b) % _mersenne_prime & _max_hash computation and the
  per-column min reduction are performed on the GPU.
- The CPU code path is preserved exactly as before, and hashvalues
  remain a NumPy uint64 array so downstream code (LSH, LSH Forest,
  LSH Ensemble, storage) is unaffected.

Includes optional tests (guarded by pytest.importorskip("cupy"))
that verify CPU and GPU produce identical hashvalues for the same
seed, num_perm, and input batches.

ekzhu commented on datasketch/minhash.py:
```diff
@@ -131,6 +139,17 @@ def _init_permutations(self, num_perm: int) -> np.ndarray:
     def _parse_hashvalues(self, hashvalues):
         return np.array(hashvalues, dtype=np.uint64)
 
+    def enable_gpu(self) -> None:
```
Can we make this a constructor parameter? What do you think?

dipeshbabu replied:
Yes, that could be much better and clean.

ekzhu commented on test/test_minhash_gpu.py:
```diff
@@ -0,0 +1,42 @@
+    return [f"token-{i}".encode("utf-8") for i in range(n)]
+
+
+class TestMinHashGPU(unittest.TestCase):
```
We may need to conditionally skip the test when the resource is not available.

ekzhu reviewed (commented):
Can we add a benchmark script to show the performance comparison -- just use randomly generated data.

dipeshbabu commented:
> Can we add a benchmark script to show the performance comparison -- just use randomly generated data.

Yes, I am adding benchmark script as well

ekzhu commented on datasketch/minhash.py:
```diff
@@ -186,10 +241,40 @@ def update_batch(self, b: Iterable) -> None:
+                self._a_gpu = cp.asarray(a, dtype=cp.uint64)
+                self._b_gpu = cp.asarray(b, dtype=cp.uint64)
+
+                hv_gpu = cp.asarray(hv_list, dtype=cp.uint64).reshape(-1, 1)
```
This line and below until return should reduce an indentation level

ekzhu commented on datasketch/minhash.py:
```diff
@@ -168,13 +217,19 @@ def _hash_32(b):
         The values will be hashed using the hash function specified by
         the `hashfunc` argument in the constructor.
 
+        If GPU is enabled globally (via constructor or enable_gpu()) or per-call
```
I don't think we should have this override, it creates too much flexibility that leads to more ways of writing the code, and more room for errors.

ekzhu commented:
@dipeshbabu , thanks for the figures! Can we update docs (https://github.com/ekzhu/datasketch/blob/master/docs/minhash.rst) and include these figures? I think we can have two plots:

1. runtime comparison for CPU vs GPU over different num_perm, with fixed n
2. runtime comparison for CPU vs GPU over different n, with fixed num_perm. 

We can do this in a separate PR.

ekzhu commented on docs/minhash.rst:
```diff
@@ -115,6 +115,41 @@ MurmurHash3: `mmh3 <https://pypi.org/project/mmh3/>`__
+    python benchmark/sketches/minhash_gpu_benchmark.py \
+      --sizes 1000 10000 50000 --num-perm 128 256 512 --repeats 5 --warmup --plot
+
+This produces CSV to stdout and figures in ``benchmark_out/``.
```
Let's skip the benchmark running code block, and include the figures directly. Also include some discussion on the results.

1. runtime comparison for CPU vs GPU over different num_perm, with fixed n
2. runtime comparison for CPU vs GPU over different n, with fixed num_perm.

dipeshbabu commented on docs/minhash.rst:
```diff
@@ -95,31 +97,84 @@ in the constructor.
+- ``'disable'`` (default): always CPU.
+- ``'detect'``: use GPU if available, otherwise fallback to CPU.
+- ``'always'``: require GPU; raises ``RuntimeError`` if no CUDA device is available.
+
```
> Add a small code snippet here showing it being used with `MinHash` constructor can help this usage get indexed by LLMs.

It's done.

dipeshbabu commented:
> @dipeshbabu thanks for the update! Can you keep only the two figures that was used in the doc?

@ekzhu Can you review it now? I updated and also added code snippet for using GPU with MinHash.
